File: src/java/tusan/nodes/base/function/FunctionNode.java
package net.tutla.tums.tusan.nodes.base.function;

import net.tutla.tums.tusan.Node;
import net.tutla.tums.tusan.Types;
import net.tutla.tums.tusan.Utils;
import net.tutla.tums.tusan.interpreter.Interpreter;
import net.tutla.tums.tusan.lexer.Token;
import net.tutla.tums.tusan.lexer.TokenType;
import net.tutla.tums.tusan.nodes.expression.Expression;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FunctionNode extends Node { // named this way to avoid any conflicts
    String name = token.value;
    public List<FunctionParameter> parameters = new ArrayList<>();
    public Interpreter functionInterpreter = new Interpreter();

    public FunctionNode(Token token){
        super(token);
    }

    public FunctionNode create(){
        Boolean paramChecks = false; // checking optional parameters

        interpreter.expectTokenType(TokenType.LEFT_PAR);
        while (interpreter.getNextToken().type == TokenType.IDENTIFIER){
            String parameterName = interpreter.nextToken().value;
            FunctionParameter parameter = new FunctionParameter(parameterName);
            if (interpreter.getNextToken().type == TokenType.COLON){
                interpreter.nextToken();
                parameter.setType(Utils.getTypeFromName(interpreter.nextToken()));
            } else {
                parameter.setType(Types.ANY);
            }

            if (interpreter.getNextToken().type == TokenType.EQUAL){
                paramChecks = true;
                interpreter.nextToken();
                parameter.setFallback(new Expression(interpreter.nextToken()).create().value);
            } else {
                if (!paramChecks){
                    parameter.setFallback(null);
                    parameter.setRequired(true);
                } else {
                    interpreter.error("SyntaxError", "You cannot use required parameters after optional parameters", Arrays.asList("Move the optional parameters to the end of the function"));
                }
            }

            parameters.add(parameter);
        }
        interpreter.expectTokenType(TokenType.RIGHT_PAR);
        interpreter.expectToken(TokenType.KEYWORD, "that");

        List<Token> tokens = new ArrayList<>();
        int structures = 0;

        Token nextToken = interpreter.getNextToken();
        if (nextToken == null){
            interpreter.error("SyntaxError", "Unexpected end of file in function definition", Arrays.asList("Make sure your function has an 'end' token"));
            return this;
        }

        if (nextToken.type == TokenType.ENDSTRUCTURE){
            interpreter.nextToken();
            tokens.add(new Token(TokenType.ENDSCRIPT, "", interpreter));
        } else {
            while (true){
                nextToken = interpreter.nextToken();
                if (nextToken == null){
                    interpreter.error("SyntaxError", "Unexpected end of file in function definition", Arrays.asList("Make sure your function has an 'end' token"));
                    return this;
                }


                if (nextToken.type == TokenType.STRUCTURE){
                    structures++;
                    tokens.add(nextToken);
                } else if (nextToken.type == TokenType.ENDSTRUCTURE){
                    if (structures == 0){
                        break;
                    } else {
                        tokens.add(nextToken);
                        structures--;
                    }
                } else {
                    tokens.add(nextToken);
                }
            }
        }
        tokens.add(new Token(TokenType.ENDSCRIPT, "function", interpreter));

        functionInterpreter.setup(interpreter.data, tokens, null, null );
        interpreter.data.funcs.put(name, new FunctionRegistry(parameters, functionInterpreter));

        return this;
    }
}

File: src/python/src/tusan/nodes/base/function/FunctionNode.py
import re
import asyncio

from tusan.Node import *
from tusan.nodes.expression.Condition import *

from tusan.Variable import *

class FunctionNode(Node):
    def __init__(self, token: Token):
        
        self.interpreter = token.interpreter
        self.token = token

    async def create(self):
        from tusan.interpreter import Interpreter
        from tusan.nodes.statement import StatementNode
        from tusan.nodes.expressions import ExpressionNode
        self.name = self.token.value
        self.params = []

        """
        Each Param will look like:

        function add num1 num2:NUMBER num3:NUMBER

        num1 will be: num1 (takes in expression)
        num2 will be: NUMBER:num2 
        """
        param_checks_begun = False
        while self.interpreter.get_next_token() and self.interpreter.get_next_token().type in ["IDENTIFIER"]:
            param = self.interpreter.next_token().value

            # Type
            if self.interpreter.get_next_token().type == "COLON":
                self.interpreter.next_token()
                formed_param = {
                    "name": param,
                    "type": self.interpreter.next_token().value
                }
            else:
                formed_param = {
                    "name": param,
                    "type": "ANY"
                }
            
            # Fallback
            if self.interpreter.is_token("COMPARISION:is"):
                param_checks_begun = True
                self.interpreter.next_token()
                formed_param["fallback"] = (await ExpressionNode(self.interpreter.next_token()).create()).value
            else:
                if not param_checks_begun:
                    formed_param["fallback"] = [None, 5678, "no chance"]
                else:
                    self.interpreter.error("SyntaxError", "You cannot use required paramaters after optional paramaters", notes=["Move the optional paramaters to the end of the function"])
            
            self.params.append(formed_param)
        


        self.interpreter.expect_token("KEYWORD:that")

        self.function_interpreter = Interpreter()

        # puts all the tokens in here
        self.tokens = []
        internal_structure_count = 0
        
        # Check if next token is end
        next_token = self.interpreter.get_next_token()
        if next_token is None:
            self.interpreter.error("SyntaxError", "Unexpected end of file in function definition", notes=["Make sure your function has an 'end' token"])
            return self
            
        if next_token.type == "ENDSTRUCTURE":
            self.interpreter.next_token() 
            self.tokens.append(Token("ENDSCRIPT", "", self.interpreter))
        else:
            while True:
                nxt_tkn = self.interpreter.get_next_token()
                if nxt_tkn is None:
                    self.interpreter.error("SyntaxError", "Unexpected end of file in function definition", notes=["Make sure your function has an 'end' token"])
                    return self
                
                if nxt_tkn.type == "STRUCTURE":
                    internal_structure_count += 1
                    tkn_to_append = self.interpreter.next_token()
                    tkn_to_append.interpreter = self.function_interpreter
                    self.tokens.append(tkn_to_append)
                elif nxt_tkn.type == "ENDSTRUCTURE":
                    if internal_structure_count == 0:
                        self.interpreter.next_token()
                        break
                    else:
                        tkn_to_append = self.interpreter.next_token()
                        tkn_to_append.interpreter = self.function_interpreter
                        self.tokens.append(tkn_to_append)
                        internal_structure_count -= 1
                else:
                    tkn_to_append = self.interpreter.next_token()
                    tkn_to_append.interpreter = self.function_interpreter
                    self.tokens.append(tkn_to_append)

            self.tokens.append(Token("ENDSCRIPT", "", self.interpreter))

        self.function_interpreter.setup(tokens=self.tokens, data=self.interpreter.data, bot=self.interpreter.bot)
        self.interpreter.data["funcs"][self.name] = {
            "params": self.params,
            "interpreter": self.function_interpreter
        }
        
        return self
                    

